### PR TITLE
board: arm: added board efm32pg_stk3401a

### LINKED ISSUE #31071: board: arm: SiliconLabs: add support to development kit efm32pg_stk3401a
**Is this request related to a missing driver support for a particular hardware platform, SoC or board? Please describe.**
I'd like to add support to the board SLSTK3401A from Silicon Labs ([SLSTK3401A](https://www.silabs.com/documents/public/user-guides/ug154-stk3401-user-guide.pdf))

**Describe why you are asking for this support?**
To perform this porting, I had to create some new files at the `./modules/hal/silabs/gecko/Device/SiliconLabs/EFM32PG1B` and `./modules/hal/silabs/gecko/emLib`

If this is a new board or SoC, please state whether you are willing to maintain the Zephyr support for it if it is included in the main tree -> YES! I can do that

**Additional context**
Board picture:
efm32pg_stk3401a

### PR DESCRIPTION
**Pull request by rdmeneze**
Signed-off-by: Rafael Dias Menezes <rdmeneze@gmail.com>

### CODE REVIEW

**Comment by rdmeneze** [comment]
what is the error "1: UC6 Body has no content, should at least have 1 line." ?

**Comment by MaureenHelm** [comment]
> what is the error "1: UC6 Body has no content, should at least have 1 line." ?

It means the commit message is missing a body after the title line, but it looks like you figured that out.

Please have a look at the failed buildkite check, there are some test applications that aren't building successfully for this new board.

Thanks for the contribution!

**Review by ioannisg — APPROVED** [review]
Looks ok to me.
You could add HW_STACK_PROTECTION=y in the defconfig (it's sort of default setting now for cortex-m boards)

File: boards/arm/efm32pg_stk3401a/Kconfig.board [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# EFM32PG STK3402A board
```
Comment (chrta) [inline boards/arm/efm32pg_stk3401a/Kconfig.board:1]: STK3402A -> STK3401A

File: boards/arm/efm32pg_stk3401a/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+- `EFM32PG Website`_
+- `EFM32PG1 Datasheet`_
+- `EFM32PG-STK3401A Website`_
+- `EFM32PG-STK3401A User Guide`_
+
+Supported Features
+==================
+
+The efm32pg_stk3402a board configuration supports the following hardware features:
```
Comment (chrta) [inline boards/arm/efm32pg_stk3401a/doc/index.rst:40]: efm32pg_stk3402a -> efm32pg_stk3401a

File: boards/arm/efm32pg_stk3401a/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+	``boards/arm/efm32pg_stk3401a/efm32pg_stk3401a_defconfig``
+
+Other hardware features are currently not supported by the port.
+
+Connections and IOs
+===================
+
+The EFM32PG1 SoC has twelve GPIO controllers (PORTA to PORTL), but only four
+are currently enabled (PORTA, PORTB, PORTD and PORTF) for the EFM32PG-STK3402A
```
Comment (chrta) [inline boards/arm/efm32pg_stk3401a/doc/index.rst:77]: STK3402A -> STK3401A

File: boards/arm/efm32pg_stk3401a/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+The default configuration can be found in the defconfig file:
+
+	``boards/arm/efm32pg_stk3401a/efm32pg_stk3401a_defconfig``
+
+Other hardware features are currently not supported by the port.
+
+Connections and IOs
+===================
+
+The EFM32PG1 SoC has twelve GPIO controllers (PORTA to PORTL), but only four
```
Comment (chrta) [inline boards/arm/efm32pg_stk3401a/doc/index.rst:76]: I think this variant has at most 6 GPIO controllers, PORTA to PORTF.

File: dts/arm/silabs/efm32_pg_1b.dtsi [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+			compatible = "silabs,gecko-usart";
+			reg = <0x40010400 0x400>;
+			interrupts = <20 0 21 0>;
+			interrupt-names = "rx", "tx";
+			peripheral-id = <1>;
+			status = "disabled";
+			label = "USART_1";
+		};
+
+		usart2: usart@40010800 { /* USART2 */
```
Comment (chrta) [inline dts/arm/silabs/efm32_pg_1b.dtsi:66]: AFAIK there are only 2 USARTS available, so usart2 and usart3 should be removed.

File: dts/arm/silabs/efm32_pg_1b.dtsi [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x4000c000 0x400>;
+			interrupts = <17 0>;
+			label = "I2C_0";
+			status = "disabled";
+		};
+
+		i2c1: i2c@4000c400 {
```
Comment (chrta) [inline dts/arm/silabs/efm32_pg_1b.dtsi:106]: There is only one i2c controller, please remove i2c1.

File: dts/arm/silabs/efm32_pg_1b.dtsi [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+				compatible = "silabs,gecko-gpio-port";
+				reg = <0x4000a0f0 0x30>;
+				peripheral-id = <5>;
+				label = "GPIO_F";
+				gpio-controller;
+				#gpio-cells = <2>;
+				status = "disabled";
+			};
+
+			gpioi: gpio@4000a180 {
```
Comment (chrta) [inline dts/arm/silabs/efm32_pg_1b.dtsi:198]: Please remove the unavailable gpio ports (gpioi and following)

File: dts/arm/silabs/efm32_pg_1b.dtsi [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+			wdog0: wdog@40052000 {
+				compatible = "silabs,gecko-wdog";
+				reg = <0x40052000 0x2C>;
+				peripheral-id = <0>;
+				label = "WDOG0";
+				interrupts = <2 0>;
+				status = "disabled";
+			};
+
+			wdog1: wdog@40052400 {
```
Comment (chrta) [inline dts/arm/silabs/efm32_pg_1b.dtsi:237]: Please remove wdog1, which is not available on this soc.

File: dts/arm/silabs/efm32_pg_1b.dtsi [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+				compatible = "silabs,gecko-wdog";
+				reg = <0x40052400 0x2C>;
+				peripheral-id = <1>;
+				label = "WDOG1";
+				interrupts = <3 0>;
+				status = "disabled";
+			};
+
+			trng0: trng@4001d000 {
```
Comment (chrta) [inline dts/arm/silabs/efm32_pg_1b.dtsi:246]: Please remove the trng, which is not available here.

File: soc/arm/silabs_exx32/efm32pg1b/Kconfig.series [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+	select SOC_FAMILY_EXX32
+	select HAS_POWER_STATE_SLEEP_1
+	select HAS_POWER_STATE_SLEEP_2
+	select HAS_POWER_STATE_SLEEP_3
+	select SOC_GECKO_HAS_INDIVIDUAL_PIN_LOCATION
+	select SOC_GECKO_HAS_HFRCO_FREQRANGE
+	select SOC_GECKO_CMU
+	select SOC_GECKO_EMU
+	select SOC_GECKO_GPIO
+	select SOC_GECKO_TRNG
```
Comment (chrta) [inline soc/arm/silabs_exx32/efm32pg1b/Kconfig.series:24]: This soc does not have a trng, please remove this line.

File: soc/arm/silabs_exx32/efm32pg1b/Kconfig.soc [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# EFM32PG12B (Pearl Gecko) MCU line
```
Comment (chrta) [inline soc/arm/silabs_exx32/efm32pg1b/Kconfig.soc:1]: EFM32PG12B -> EFM32PG1B

File: soc/arm/silabs_exx32/efm32pg1b/soc.h [hunk 12]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (c) 2017 Christian Taedcke
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Board configuration macros for the efm32pg12b soc
```
Comment (chrta) [inline soc/arm/silabs_exx32/efm32pg1b/soc.h:9]: efm32pg12b -> efm32pg1b

**Review by chrta — CHANGES_REQUESTED** [review]
Thank you for your contribution. See my comments.

1. I have not checked the register addresses and pins, yet. So this is still pending.
2. Please the  adapt the hal_silabs revision in west.yml https://github.com/zephyrproject-rtos/zephyr/blob/master/west.yml#L71 so it pulls in the revision with your changes from https://github.com/zephyrproject-rtos/hal_silabs/pull/10 See also https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-module

```
    - name: hal_silabs
      revision: pull/10/head
      path: modules/hal/silabs
```

**Comment by chrta** [comment]
@rdmeneze Unfortunately, i do not have the board here to test this. To verify that it works as expected, can you execute twister (was sanitycheck) and perform device-tests? This way you should easily see if something unexpected fails.
See also https://docs.zephyrproject.org/latest/guides/test/twister.html#executing-tests-on-a-single-device
```
$ ./scripts/twister -p efm32pg_stk3401a -v --device-testing --device-serial /dev/ttyACM0
```

File: boards/arm/efm32pg_stk3401a/doc/index.rst (hunk 4 above)
Comment (rdmeneze) [reply boards/arm/efm32pg_stk3401a/doc/index.rst:76]: This variant has 5 ports: PORTA-PORTD and PORTF

**Comment by rdmeneze** [comment]
> Unfortunately, i do not have the board here to test this. To verify that it works as expected, can you execute twister (was sanitycheck) and perform device-tests?

I don't have enough disk space to perform this test. I'll try uninstall some applications to perform these tests

**Review by rdmeneze — COMMENTED** [review]
done

**Comment by rdmeneze** [comment]
> Unfortunately, i do not have the board here to test this. To verify that it works as expected, can you execute twister (was sanitycheck) and perform device-tests?

this is the first time that I do this. 

The report is below: 

> Hardware distribution summary:
> | Board            | ID   |   Counter |
> |------------------|------|-----------|
> | efm32pg_stk3401a |      |       174 |
> INFO    - Saving reports...
> INFO    - Writing xunit report /home/rdias/sources/zephyr_blink_efm32/zephyr/twister-out/twister.xml...
> INFO    - Writing xunit report /home/rdias/sources/zephyr_blink_efm32/zephyr/twister-out/twister_report.xml...
> INFO    - Writing JSON report /home/rdias/sources/zephyr_blink_efm32/zephyr/twister-out/twister.json
> INFO    - Run completed


Tests results:
>                 {
>                     "testcase":"arch.interrupt.gen_isr_table.arm_mainline.build_time_direct_interrupt",
>                     "arch":"arm",
>                     "type":"integration",
>                     "platform":"efm32pg_stk3401a",
>                     "status":"failed",
>                     "reason":"Failed",
>                     "execution_time":4.008305549621582,
>                     "test_output":"*** Booting Zephyr OS build zephyr-v2.4.0-2712-gea1fc5e48235  ***\r\nSTART - Test gen_isr_tables\r\nIRQ configuration (total lines 50):\r\nRunning test suite context\r\n===================================================================\r\nSTART - test_build_time_direct_interrupt\r\nisr1 isr=0x491 irq=49\r\nisr2 isr=0x4fd irq=48\r\nChecking _irq_vector_table entry 49 for irq 49\r\ntriggering irq 49\r\ninterrupt 49 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:272: test_build_time_direct_interrupt: (check_vector(isr1, ISR1_OFFSET) is non-zero)\r\ncheck direct interrpt isr1 failed\r\nFAIL - test_build_time_direct_interrupt\r\n===================================================================\r\nSTART - test_build_time_interrupt\r\n_sw_isr_table at location 0x20000088\r\nisr3 isr=0x401 irq=47 param=0xb01dface\r\nChecking _sw_isr_table entry 47 for irq 47\r\ntriggering irq 47\r\ninterrupt 47 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:308: test_build_time_interrupt: (check_sw_isr(isr3, ISR3_ARG, ISR3_OFFSET) is non-zero)\r\ncheck interrupt isr3 failed\r\nFAIL - test_build_time_interrupt\r\n===================================================================\r\nSTART - test_run_time_interrupt\r\nisr5 isr=0x449 irq=45 param=0xf0ccac1a\r\nChecking _sw_isr_table entry 45 for irq 45\r\ntriggering irq 45\r\ninterrupt 45 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:350: test_run_time_interrupt: (check_sw_isr(isr5, ISR5_ARG, ISR5_OFFSET) is non-zero)\r\ntest dynamic interrupt isr5 failed\r\nFAIL - test_run_time_interrupt\r\n===================================================================\r\nTest suite context failed.\r\n===================================================================\r\nPROJECT EXECUTION FAILED\r\n"
>                 },
>                 {
>                     "testcase":"arch.interrupt.gen_isr_table.arm_mainline.build_time_interrupt",
>                     "arch":"arm",
>                     "type":"integration",
>                     "platform":"efm32pg_stk3401a",
>                     "status":"failed",
>                     "reason":"Failed",
>                     "execution_time":4.008305549621582,
>                     "test_output":"*** Booting Zephyr OS build zephyr-v2.4.0-2712-gea1fc5e48235  ***\r\nSTART - Test gen_isr_tables\r\nIRQ configuration (total lines 50):\r\nRunning test suite context\r\n===================================================================\r\nSTART - test_build_time_direct_interrupt\r\nisr1 isr=0x491 irq=49\r\nisr2 isr=0x4fd irq=48\r\nChecking _irq_vector_table entry 49 for irq 49\r\ntriggering irq 49\r\ninterrupt 49 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:272: test_build_time_direct_interrupt: (check_vector(isr1, ISR1_OFFSET) is non-zero)\r\ncheck direct interrpt isr1 failed\r\nFAIL - test_build_time_direct_interrupt\r\n===================================================================\r\nSTART - test_build_time_interrupt\r\n_sw_isr_table at location 0x20000088\r\nisr3 isr=0x401 irq=47 param=0xb01dface\r\nChecking _sw_isr_table entry 47 for irq 47\r\ntriggering irq 47\r\ninterrupt 47 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:308: test_build_time_interrupt: (check_sw_isr(isr3, ISR3_ARG, ISR3_OFFSET) is non-zero)\r\ncheck interrupt isr3 failed\r\nFAIL - test_build_time_interrupt\r\n===================================================================\r\nSTART - test_run_time_interrupt\r\nisr5 isr=0x449 irq=45 param=0xf0ccac1a\r\nChecking _sw_isr_table entry 45 for irq 45\r\ntriggering irq 45\r\ninterrupt 45 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:350: test_run_time_interrupt: (check_sw_isr(isr5, ISR5_ARG, ISR5_OFFSET) is non-zero)\r\ntest dynamic interrupt isr5 failed\r\nFAIL - test_run_time_interrupt\r\n===================================================================\r\nTest suite context failed.\r\n===================================================================\r\nPROJECT EXECUTION FAILED\r\n"
>                 },
>                 {
>                     "testcase":"arch.interrupt.gen_isr_table.arm_mainline.run_time_interrupt",
>                     "arch":"arm",
>                     "type":"integration",
>                     "platform":"efm32pg_stk3401a",
>                     "status":"failed",
>                     "reason":"Failed",
>                     "execution_time":4.008305549621582,
>                     "test_output":"*** Booting Zephyr OS build zephyr-v2.4.0-2712-gea1fc5e48235  ***\r\nSTART - Test gen_isr_tables\r\nIRQ configuration (total lines 50):\r\nRunning test suite context\r\n===================================================================\r\nSTART - test_build_time_direct_interrupt\r\nisr1 isr=0x491 irq=49\r\nisr2 isr=0x4fd irq=48\r\nChecking _irq_vector_table entry 49 for irq 49\r\ntriggering irq 49\r\ninterrupt 49 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:272: test_build_time_direct_interrupt: (check_vector(isr1, ISR1_OFFSET) is non-zero)\r\ncheck direct interrpt isr1 failed\r\nFAIL - test_build_time_direct_interrupt\r\n===================================================================\r\nSTART - test_build_time_interrupt\r\n_sw_isr_table at location 0x20000088\r\nisr3 isr=0x401 irq=47 param=0xb01dface\r\nChecking _sw_isr_table entry 47 for irq 47\r\ntriggering irq 47\r\ninterrupt 47 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:308: test_build_time_interrupt: (check_sw_isr(isr3, ISR3_ARG, ISR3_OFFSET) is non-zero)\r\ncheck interrupt isr3 failed\r\nFAIL - test_build_time_interrupt\r\n===================================================================\r\nSTART - test_run_time_interrupt\r\nisr5 isr=0x449 irq=45 param=0xf0ccac1a\r\nChecking _sw_isr_table entry 45 for irq 45\r\ntriggering irq 45\r\ninterrupt 45 didn't run once, ran 0 times\r\nAssertion failed at WEST_TOPDIR/zephyr/tests/kernel/gen_isr_table/src/main.c:350: test_run_time_interrupt: (check_sw_isr(isr5, ISR5_ARG, ISR5_OFFSET) is non-zero)\r\ntest dynamic interrupt isr5 failed\r\nFAIL - test_run_time_interrupt\r\n===================================================================\r\nTest suite context failed.\r\n===================================================================\r\nPROJECT EXECUTION FAILED\r\n"
>                 },

**Review by chrta — CHANGES_REQUESTED** [review]
Thank you for the fixes so far. I now reviewed everything. A few irq numbers are one off (EFM32PG1B vs EFM3212B). I also discovered issues that are present in the other efm32 device trees. I will have a look at them later on.

After fixing my comments, could you please restructure you commits and change commit messages? I think https://github.com/zephyrproject-rtos/zephyr/pull/22423/commits is a good example. So two commits are fine. The first one adding support for the soc (soc, dts and west.yml change), the second one for the board.

**Comment by chrta** [comment]
@rdmeneze Regarding your twister run:

Normally you should see the execution of around 200 tests on your console with a result of either PASSED or FAILED. E.g. when i run twister for my efm32pg_stk3402a board (you must have JLink in your PATH, so programming works):

```
$ scripts/twister -p efm32pg_stk3402a -v --device-testing --device-serial /dev/ttyACM0
```

I get the following result:
```
ZEPHYR_BASE unset, using "/home/user/projects/zephyrproject/zephyr"
Renaming output directory to /home/user/projects/zephyrproject/zephyr/twister-out.5
INFO    - Zephyr version: zephyr-v2.4.0-2814-g8a80eb99620c
INFO    - JOBS: 6
INFO    - Building initial testcase list...
INFO    - 1058 test suites (1030 configurations) selected, 794 configurations discarded due to filters.

Device testing on:

| Platform         | ID   | Serial device   |
|------------------|------|-----------------|
| efm32pg_stk3402a |      | /dev/ttyACM0    |

INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    -   1/236 efm32pg_stk3402a          samples/synchronization/sample.kernel.synchronization PASSED (device 3.247s)
INFO    -   2/236 efm32pg_stk3402a          samples/userspace/hello_world_user/sample.helloworld PASSED (device 4.140s)
INFO    -   3/236 efm32pg_stk3402a          samples/smp/pktqueue/sample.smp.pktqueue           SKIPPED (filter)
INFO    -   4/236 efm32pg_stk3402a          samples/smp/pi/sample.smp.pi                       SKIPPED (filter)
INFO    -   5/236 efm32pg_stk3402a          samples/userspace/prod_consumer/sample.userspace.prod_consumer PASSED (device 5.624s)
INFO    -   6/236 efm32pg_stk3402a          samples/userspace/shared_mem/sample.kernel.memory_protection.shared_mem PASSED (device 4.007s)
...
INFO    - 235/236 efm32pg_stk3402a          tests/application_development/cpp/application_development.cpp.main PASSED (device 3.308s)
INFO    - 236/236 efm32pg_stk3402a          tests/application_development/gen_inc_file/buildsystem.include_file PASSED (device 3.367s)

INFO    - 217 of 221 test configurations passed (98.19%), 4 failed, 809 skipped with 0 warnings in 1523.68 seconds
INFO    - In total 4849 test cases were executed, 4344 skipped on 1 out of total 321 platforms (0.31%)
INFO    - 221 test configurations executed on platforms, 0 test configurations were only built.

Hardware distribution summary:

| Board            | ID   |   Counter |
|------------------|------|-----------|
| efm32pg_stk3402a |      |       221 |
INFO    - Saving reports...
INFO    - Writing xunit report /home/chrta/projects/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/chrta/projects/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Writing JSON report /home/chrta/projects/zephyrproject/zephyr/twister-out/twister.json
INFO    - Run completed
```

When you run `scripts/twister -p efm32pg_stk3401a -v --device-testing --device-serial /dev/ttyACM0` i would also expect about 4 failed tests and about 200 passed.

File: dts/arm/silabs/efm32_pg_1b.dtsi [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+			interrupt-names = "rx", "tx";
+			peripheral-id = <1>;
+			status = "disabled";
+			label = "USART_1";
+		};
+
+		leuart0: leuart@4004a000 { /* LEUART0 */
+			compatible = "silabs,gecko-leuart";
+			reg = <0x4004a000 0x400>;
+			interrupts = <22 0>;
```
[the remark this answers is not in the document]
Comment (chrta) [reply dts/arm/silabs/efm32_pg_1b.dtsi:68]: This is still open

File: soc/arm/silabs_exx32/efm32pg1b/Kconfig.defconfig.series [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+config SOC_SERIES
+	default "efm32pg1b"
+
+config SOC_PART_NUMBER
+	default "EFM32PG1B200F256GM48" if SOC_PART_NUMBER_EFM32PG1B200F256GM48
+
+config NUM_IRQS
+	# must be >= the highest interrupt number used
+	default 30
```
Comment (chrta) [inline soc/arm/silabs_exx32/efm32pg1b/Kconfig.defconfig.series:16]: Interrupt FPUEH is IRQ number 33, so this should be 33 or 34 (i am not sure which one). For efm32pg12 is is one above FPUEH. So 34 would be the equivalent value here.

**Review by chrta — CHANGES_REQUESTED** [review]
Most of the comments are fixed, only 2 are open (one new and one from the previous review that you probably missed).
Changing the commits is also open.
Otherwise this looks good.

**Review by chrta — APPROVED** [review]
LGTM. Unfortunately i could not test this on hardware, since currently i do not have access to this board.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**